Unable to use auto_impl with async-trait

I was trying to use `auto_impl` to delegate the implementation of `Box<T> where T: Trait` and `Trait` was a trait annotated with the [`async-trait`](https://github.com/dtolnay/async-trait) procedural macro. I was using it like this:

```rust
#[async_trait]
#[auto_impl(Box)]
trait Foo: Sync {
    async fn bar(&self) -> Result<u32, Error>;
}
```

The ordering of the above macros matters, as it ensures that `async_trait` transforms `Foo` into the following Rust code before `auto_impl` parses it:

```rust
#[auto_impl(Box)]
trait Foo: Sync {
    fn bar<'life0, 'async_trait>(
        &'life0 self
    ) -> Pin<Box<dyn Future<Output = Result<u32, Error>> + Send + 'async_trait>> where
        'life0: 'async_trait,
        Self: 'async_trait;
}
```

Provided that the `Sync` bound shown above is present as a supertrait of `Foo`, the above code should be object safe. I expected `#[auto_impl(Box)]` to emit the following implementation:

```rust
impl<T: Foo + ?Sized> Foo for Box<T>
where
    Box<T>: Sync
```

However, it emitted the following implementation instead:

```rust
impl<T: Foo> Foo for Box<T>
where
    Box<T>: Sync
```

This meant that the following code was able to compile with a manual `Box` implementation but failed under the `auto_impl`-generated implementation:

```rust
fn assert_impl<T: Foo>() {}

fn foo() {
    assert_impl::<Box<dyn Foo>>(); // ERROR: Not object safe with `auto_impl`
}
```

Do you have an idea as to why the `?Sized` bound was not applied in this case?

See https://github.com/ebkalderon/tower-lsp/pull/109 for the original pull request that inspired this issue.

Hmm, I've revised the commit on the [`use-auto-impl-box`](https://github.com/ebkalderon/tower-lsp/tree/use-auto-impl-box) branch of `tower-lsp` to test out `master` of `auto_impl`, and it's failing to compile with the following error:

```
error[E0261]: use of undeclared lifetime name `'async_trait`
  --> src/lib.rs:98:1
   |
98 | #[auto_impl(Box)]
   | ^^^^^^^^^^^^^^^^^ undeclared lifetime

error: aborting due to previous error

For more information about this error, try `rustc --explain E0261`.
error: could not compile `tower-lsp`.
```

Awesome work, @LukasKalbertodt! That's a very subtle change, but it makes perfect sense. I really appreciate you looking into it.

I tried enabling it in `tower-lsp` by adding `#[auto_impl(Box)]` underneath the `#[async_trait]` attribute, but now I'm running into another issue where this static assertion nonetheless fails to compile:

```rust
fn _assert_object_safe() {
    fn assert_impl<T: LanguageServer>() {}
    assert_impl::<Box<dyn LanguageServer>>();
}
```

Error message:

```
error[E0277]: the size for values of type `(dyn LanguageServer + 'static)` cannot be known at compilation time
   --> src/lib.rs:748:5
    |
747 |     fn assert_impl<T: LanguageServer>() {}
    |                       -------------- required by this bound in `_assert_object_safe::assert_impl`
748 |     assert_impl::<Box<dyn LanguageServer>>();
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `std::marker::Sized` is not implemented for `(dyn LanguageServer + 'static)`
    = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
    = note: required because of the requirements on the impl of `LanguageServer` for `std::boxed::Box<(dyn LanguageServer + 'static)>
```

This seems to be because `auto_impl(Box)` does not include a `T: ?Sized` bound on `T`, which my current hand-written implementation does. I was under impression the changes in auto-impl-rs/auto_impl#52 were supposed to handle this.

Any idea why the `?Sized` bound isn't being added here?

--- a/src/gen.rs
+++ b/src/gen.rs
@@ -185,6 +185,19 @@ fn gen_header(
                         }
                     })
                     .flat_map(|p| &p.bounds)
+                    .filter_map(|b| {
+                        // We are only interested in trait bounds. That's
+                        // because lifetime bounds on `Self` do not need to be
+                        // added to the impl header. That's because all values
+                        // "derived" from `self` also meet the same lifetime
+                        // bound as `self`. In simpler terms: while `self.field`
+                        // might not be `Clone` even if `Self: Clone`,
+                        // `self.field` is always `: 'a` if `Self: 'a`.
+                        match b {
+                            TypeParamBound::Trait(t) => Some(t),
+                            TypeParamBound::Lifetime(_) => None,
+                        }
+                    })
             });
 
         // Determine if our proxy type needs a lifetime parameter

--- a/tests/compile-fail/attr_on_enum.stderr
+++ b/tests/compile-fail/attr_on_enum.stderr
@@ -7,3 +7,5 @@ error: couldn't parse trait item
   |
 4 | #[auto_impl(&, &mut)]
   | ^^^^^^^^^^^^^^^^^^^^^
+  |
+  = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

--- a/tests/compile-fail/attr_on_fn.stderr
+++ b/tests/compile-fail/attr_on_fn.stderr
@@ -7,3 +7,5 @@ error: couldn't parse trait item
   |
 4 | #[auto_impl(&, &mut)]
   | ^^^^^^^^^^^^^^^^^^^^^
+  |
+  = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

--- a/tests/compile-fail/attr_on_impl_block.stderr
+++ b/tests/compile-fail/attr_on_impl_block.stderr
@@ -7,3 +7,5 @@ error: couldn't parse trait item
   |
 6 | #[auto_impl(&, &mut)]
   | ^^^^^^^^^^^^^^^^^^^^^
+  |
+  = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

--- a/tests/compile-fail/attr_on_struct.stderr
+++ b/tests/compile-fail/attr_on_struct.stderr
@@ -7,3 +7,5 @@ error: couldn't parse trait item
   |
 4 | #[auto_impl(&, &mut)]
   | ^^^^^^^^^^^^^^^^^^^^^
+  |
+  = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

--- a/tests/compile-fail/attr_on_type.stderr
+++ b/tests/compile-fail/attr_on_type.stderr
@@ -7,3 +7,5 @@ error: couldn't parse trait item
   |
 4 | #[auto_impl(&, &mut)]
   | ^^^^^^^^^^^^^^^^^^^^^
+  |
+  = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

--- a/tests/compile-fail/attr_on_unit_struct.stderr
+++ b/tests/compile-fail/attr_on_unit_struct.stderr
@@ -7,3 +7,5 @@ error: couldn't parse trait item
   |
 4 | #[auto_impl(&, &mut)]
   | ^^^^^^^^^^^^^^^^^^^^^
+  |
+  = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

--- a/tests/compile-fail/super_trait_not_implemented.stderr
+++ b/tests/compile-fail/super_trait_not_implemented.stderr
@@ -2,7 +2,7 @@ error[E0277]: the trait bound `std::boxed::Box<Dog>: Supi` is not satisfied
   --> $DIR/super_trait_not_implemented.rs:18:18
    |
 14 | fn requires_foo<T: Foo>(_: T) {}
-   |    ------------    --- required by this bound in `requires_foo`
+   |                    --- required by this bound in `requires_foo`
 ...
 18 |     requires_foo(Box::new(Dog)); // shouldn't, because `Box<Dog>: Supi` is not satisfied
    |                  ^^^^^^^^^^^^^ the trait `Supi` is not implemented for `std::boxed::Box<Dog>`

--- a/tests/compile-fail/trait_obj_value_self.stderr
+++ b/tests/compile-fail/trait_obj_value_self.stderr
@@ -2,7 +2,7 @@ error[E0277]: the size for values of type `dyn Trait` cannot be known at compila
   --> $DIR/trait_obj_value_self.rs:12:5
    |
 9  | fn assert_impl<T: Trait>() {}
-   |    -----------    ----- required by this bound in `assert_impl`
+   |                   ----- required by this bound in `assert_impl`
 ...
 12 |     assert_impl::<Box<dyn Trait>>();
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time

